### PR TITLE
feat(tree-view): add ability to disable nodes

### PR DESCRIPTION
**Pull request by valentin-mladenov**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Tree nodes can't be disabled.

Issue Number: CDE-1054

## What is the new behavior?

Tree nodes can be disabled when `[clrDisable]` is set to `true`.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: .storybook/helpers/files.data.ts [hunk 1]
```diff
@@ -50,15 +72,30 @@ export const filesRoot: File[] = [
...
   },
 ];
 
-export function getFileTreeNodeMarkup(files: File[] = filesRoot) {
+export function getFileTreeNodeMarkup(files: File[] = filesRoot, useSelected?: boolean) {
   return files
-    .map(file => `<clr-tree-node>${file.name}${file.files ? getFileTreeNodeMarkup(file.files) : ''}</clr-tree-node>`)
+    .map(file => {
+      let selected = '';
+      if (useSelected !== undefined && file.selected !== undefined) {
```
Comment (kevinbuhmann) [inline .storybook/helpers/files.data.ts:88]: Why does this use the selected value if `useSelected` is false?

Comment (valentin-mladenov) [reply .storybook/helpers/files.data.ts:88]: For `clrSelected` on tree view we have 3 values `true`, `false` and `undefined` (not provided).
When is `undefined` the tree view is rendered as list. Basically no checkboxes.
When `clrSelected` have value `true` or `false` tree view have selection capabilities, through checkboxes.

Comment (kevinbuhmann) [reply .storybook/helpers/files.data.ts:88]: Okay, I would rename it to `clrSelected` to match the story.

Comment (valentin-mladenov) [reply .storybook/helpers/files.data.ts:88]: I might misunderstood the question. 
Let me check very quickly, but yeah renaming it to `clrSelected` makes much more sense.

Comment (valentin-mladenov) [reply .storybook/helpers/files.data.ts]: variable is renamed in commit 9e7f698

Comment (kevinbuhmann) [reply .storybook/helpers/files.data.ts]: Thanks!

File: .storybook/helpers/files.data.ts [hunk 2]
```diff
@@ -50,15 +72,30 @@ export const filesRoot: File[] = [
...
 
-export function getFileTreeNodeMarkup(files: File[] = filesRoot) {
+export function getFileTreeNodeMarkup(files: File[] = filesRoot, clrSelected?: boolean) {
   return files
-    .map(file => `<clr-tree-node>${file.name}${file.files ? getFileTreeNodeMarkup(file.files) : ''}</clr-tree-node>`)
+    .map(file => {
+      let selected = '';
+      if (clrSelected !== undefined && file.selected !== undefined) {
+        selected = `[clrSelected]="${file.selected}"`;
+      }
```
Comment (kevinbuhmann) [inline .storybook/helpers/files.data.ts:90]:
```suggestion
      const selected = clrSelected !== undefined && file.selected !== undefined ? `[clrSelected]="${file.selected}"` : '';
```

File: .storybook/helpers/files.data.ts [hunk 3]
```diff
@@ -50,15 +72,30 @@ export const filesRoot: File[] = [
...
+
+      return (
+        `<clr-tree-node ` +
+        selected +
+        `[clrDisabled]="${file.disabled}" [clrExpanded]="${file.expanded}">${file.name}${
+          file.files ? getFileTreeNodeMarkup(file.files, clrSelected) : ''
+        }</clr-tree-node>`
+      );
```
Comment (kevinbuhmann) [inline .storybook/helpers/files.data.ts:98]:
```suggestion
      return `
        <clr-tree-node
          ${selected}
          [clrDisabled]="${file.disabled}"
          [clrExpanded]="${file.expanded}"
        >
          ${file.name}
          ${file.files ? getFileTreeNodeMarkup(file.files, clrSelected) : ''}
        </clr-tree-node>
      `;
```

Use string interpolation consistently and reformat.

File: projects/angular/src/data/tree-view/tree-node.ts [hunk 4]
```diff
@@ -118,6 +119,17 @@ export class ClrTreeNode<T> implements OnInit, AfterContentInit, AfterViewInit,
...
   }
 
+  @Input('clrDisabled')
+  get disabled(): boolean {
+    return this._model.disabled;
+  }
+  set disabled(value: boolean) {
+    console.log(value);
+    console.log(this._model.parent);
+    console.log(this._model.parent?.disabled);
```
Comment (kevinbuhmann) [inline projects/angular/src/data/tree-view/tree-node.ts:129]: These should be removed.

**Review by kevinbuhmann — CHANGES_REQUESTED** [review]

File: projects/angular/src/data/tree-view/tree-node.ts (hunk 4 above)
Comment (valentin-mladenov) [reply projects/angular/src/data/tree-view/tree-node.ts]: Removed.

File: .storybook/helpers/files.data.ts (hunk 2 above)
Comment (valentin-mladenov) [reply .storybook/helpers/files.data.ts]: Entire selected functionality is removed, because make no sense in storybook anymore.

File: .storybook/helpers/files.data.ts (hunk 3 above)
Comment (valentin-mladenov) [reply .storybook/helpers/files.data.ts]: Entire selected functionality is removed, because make no sense in storybook anymore.

File: .storybook/helpers/files.data.ts [hunk 5]
```diff
@@ -59,6 +74,12 @@ export const filesRoot: File[] = [
...
 export function getFileTreeNodeMarkup(files: File[] = filesRoot) {
   return files
-    .map(file => `<clr-tree-node>${file.name}${file.files ? getFileTreeNodeMarkup(file.files) : ''}</clr-tree-node>`)
+    .map(file => {
+      return `
+        <clr-tree-node [clrDisabled]="${file.disabled}" [clrExpanded]="${file.expanded}">
+          ${file.name}
+          ${file.files ? getFileTreeNodeMarkup(file.files) : ''}
+        </clr-tree-node>`;
+    })
```
Comment (kevinbuhmann) [inline .storybook/helpers/files.data.ts:83]:
```suggestion
    .map(file => `
      <clr-tree-node [clrDisabled]="${file.disabled}" [clrExpanded]="${file.expanded}">
        ${file.name}
        ${file.files ? getFileTreeNodeMarkup(file.files) : ''}
      </clr-tree-node>
    `)
```

The `return` keyword is not needed.

Comment (valentin-mladenov) [reply .storybook/helpers/files.data.ts]: removed in commit 795ffad

File: projects/angular/src/data/tree-view/tree-focus-manager.service.ts [hunk 6]
```diff
@@ -101,6 +101,7 @@ export class TreeFocusManagerService<T> {
     } else if (selfIndex === siblings.length - 1) {
       return this.findNextFocusable(model.parent);
     }
+
```
Comment (kevinbuhmann) [inline projects/angular/src/data/tree-view/tree-focus-manager.service.ts:104]:
```suggestion
```

The only change in this file is adding a blank line.

File: projects/angular/src/data/tree-view/tree-node.ts [hunk 7]
```diff
@@ -61,6 +61,7 @@ const TREE_TYPE_AHEAD_TIMEOUT = 200;
   ],
   host: {
     '[class.clr-tree-node]': 'true',
+    '[class.disabled]': 'this._model.disabled',
```
Comment (kevinbuhmann) [inline projects/angular/src/data/tree-view/tree-node.ts:64]: We can add `@HostBinding('class.disabled')` property instead.

**Review by kevinbuhmann — CHANGES_REQUESTED** [review]
I don't think it makes sense for disabled nodes to be styled differently when there is not a checkbox because the disabled state is not meaningful without a checkbox. It is actually visually misleading since the node is interactive.

Screenshot 2024-02-13 at 12 38 10 PM

Similarly, even with a checkbox, I don't think the caret icon should be a different color because the expand/collapse functionality is not disabled. This is also visually misleading.

Screenshot 2024-02-13 at 12 40 43 PM

Since the disabled state is only related to the section, I think it would be much better to remove the `clrDisabled` input and instead add a `ClrSelectedState.DISABLED` option for `clrSelected`.

File: projects/angular/src/data/tree-view/tree-focus-manager.service.ts (hunk 6 above)
Comment (valentin-mladenov) [reply projects/angular/src/data/tree-view/tree-focus-manager.service.ts]: removed

**Comment by kevinbuhmann** [comment]
>The client application will ultimately decide how the disabled state of tree-view will behave.
>Depending on the need a disabled node might be expanded or collapsed. Interaction is best leaved to client app interpretation.

That is not true. A disabled node is expandable and collapsible by the user.

File: .storybook/helpers/files.data.ts [hunk 8]
```diff
@@ -50,15 +70,26 @@ export const filesRoot: File[] = [
...
+    disabled: true,
     name: 'package.json',
   },
   {
     name: 'tsconfig.json',
   },
 ];
 
-export function getFileTreeNodeMarkup(files: File[] = filesRoot) {
+export function getFileTreeNodeMarkup(files: File[] = filesRoot, args = {}) {
```
Comment (kevinbuhmann) [inline .storybook/helpers/files.data.ts:81]:
```suggestion
export function getFileTreeNodeMarkup(files: File[] = filesRoot, args: { asLink?: boolean, clrSelected?: ClrSelectedState } = {}) {
```

Missing type for `args`.

File: .storybook/helpers/files.data.ts [hunk 9]
```diff
@@ -50,15 +70,26 @@ export const filesRoot: File[] = [
...
   return files
-    .map(file => `<clr-tree-node>${file.name}${file.files ? getFileTreeNodeMarkup(file.files) : ''}</clr-tree-node>`)
+    .map(
+      file => `
+        <clr-tree-node
+        [clrDisabled]="${file.disabled}"
+        [clrExpanded]="${file.expanded}"
+        ${args['clrSelected'] !== undefined && file.selected !== undefined ? `[clrSelected]="${file.selected}"` : ''}
```
Comment (kevinbuhmann) [inline .storybook/helpers/files.data.ts:88]:
```suggestion
        ${args.clrSelected !== undefined && file.selected !== undefined ? `[clrSelected]="${file.selected}"` : ''}
```

No need for string index access.

File: .storybook/helpers/files.data.ts [hunk 10]
```diff
@@ -50,15 +70,26 @@ export const filesRoot: File[] = [
...
+        >
+            ${args['asLink'] ? `<a href="javascript:;" class="clr-treenode-link">${file.name}</a>` : file.name}
```
Comment (kevinbuhmann) [inline .storybook/helpers/files.data.ts:90]:
```suggestion
            ${args.asLink ? `<a href="javascript:;" class="clr-treenode-link">${file.name}</a>` : file.name}
```

No need for string index access.

File: .storybook/stories/tree/tree.stories.ts [hunk 11]
```diff
@@ -14,7 +14,7 @@ import { setupStorybook } from '../../helpers/setup-storybook.helpers';
 const defaultStory: Story = args => ({
   template: `
     <clr-tree>
-      ${getFileTreeNodeMarkup(filesRoot)}
+      ${getFileTreeNodeMarkup(filesRoot, { asLink: args.asLink })}
```
Comment (kevinbuhmann) [inline .storybook/stories/tree/tree.stories.ts:17]:
```suggestion
      ${getFileTreeNodeMarkup(filesRoot, args)}
```

Why not just pass the whole `args` object? (Same thing the tree node story.)

File: .storybook/stories/tree/tree.stories.ts [hunk 12]
```diff
@@ -26,9 +26,10 @@ const defaultParameters: Parameters = {
   argTypes: {
     // inputs
     clrLazy: { control: { disable: true } },
+    asLink: { defaultValue: false, control: { type: 'boolean' } },
```
Comment (kevinbuhmann) [inline .storybook/stories/tree/tree.stories.ts:29]:
```suggestion
    // story helpers
    asLink: { defaultValue: false, control: { type: 'boolean' } },
```

Separate the story helpers from the component inputs.

File: projects/angular/src/data/tree-view/models/tree-node.model.ts [hunk 13]
```diff
@@ -46,21 +47,32 @@ export abstract class TreeNodeModel<T> {
...
+
   // Propagate by default when eager, don't propagate in the lazy-loaded tree.
   setSelected(state: ClrSelectedState, propagateUp: boolean, propagateDown: boolean) {
     if (state === this.selected.value) {
       return;
     }
     this.selected.next(state);
     if (propagateDown && state !== ClrSelectedState.INDETERMINATE && this.children) {
-      this.children.forEach(child => child.setSelected(state, false, true));
+      this.children.forEach(child => child.parentSetSelected(state, false, true));
```
Comment (kevinbuhmann) [inline projects/angular/src/data/tree-view/models/tree-node.model.ts:64]:
```suggestion
      this.children.forEach(child => {
        if (!child.disabled) {
          child.setSelected(state, false, true);
        }
      });
```

I would remove the `parentSetSelected` method. The name is confusing, and I don't think the method is needed.

Comment (kevinbuhmann) [reply projects/angular/src/data/tree-view/models/tree-node.model.ts:64]: Why doesn't `setSelected` check for disabled? I feel like it should, but I have not tested it. If `setSelected` checked `disabled` like the other methods, I don't think this line would need any changes at all.

**Review by kevinbuhmann — CHANGES_REQUESTED** [review]
See comments. Also, the `clrDisabled` input has no effect in the tree node story.

**Comment by kevinbuhmann** [comment]
> Also, the `clrDisabled` input has no effect in the tree node story.

I think we should merge #1298 and then add variants to the tree node story for disabled nodes.

File: .storybook/stories/tree/tree.stories.ts (hunk 12 above)
Comment (valentin-mladenov) [reply .storybook/stories/tree/tree.stories.ts:30]: corrected in commit e22dd2e

File: .storybook/stories/tree/tree.stories.ts (hunk 11 above)
Comment (valentin-mladenov) [reply .storybook/stories/tree/tree.stories.ts]: corrected in commit e22dd2e

File: .storybook/helpers/files.data.ts (hunk 10 above)
Comment (valentin-mladenov) [reply .storybook/helpers/files.data.ts]: corrected in commit e22dd2e

File: .storybook/helpers/files.data.ts (hunk 9 above)
Comment (valentin-mladenov) [reply .storybook/helpers/files.data.ts]: corrected in commit e22dd2e

File: .storybook/helpers/files.data.ts (hunk 8 above)
Comment (valentin-mladenov) [reply .storybook/helpers/files.data.ts]: corrected in commit e22dd2e

File: projects/angular/src/data/tree-view/models/tree-node.model.ts (hunk 13 above)
Comment (valentin-mladenov) [reply projects/angular/src/data/tree-view/models/tree-node.model.ts]: Function `parentSetSelected` is removed in commit  6846f60.

`setSelected` function is 'internal' for `[clrSelected]` directive. Selection through it should not be blocked, but only through user input.
TLDR is node with `[clrSelected]=true` and `[clrDisabled]=true` will not be selected.

**Comment by valentin-mladenov** [comment]
@kevinbuhmann to your main PR comment directly.

Regarding the `clrDisabled` input in stroybook that have no effect in the tree node story. I added disable children functionality.

File: .storybook/helpers/files.data.ts [hunk 14]
```diff
@@ -50,15 +70,29 @@ export const filesRoot: File[] = [
...
+    .map(
+      file => `
+        <clr-tree-node
+        [clrDisabled]="${file.disabled}"
+        [clrExpanded]="${file.expanded}"
+        ${args.clrSelected !== undefined && file.selected !== undefined ? `[clrSelected]="${file.selected}"` : ''}
+        >
+            ${args.asLink ? `<a href="javascript:;" class="clr-treenode-link">${file.name}</a>` : file.name}
+          ${file.files ? getFileTreeNodeMarkup(file.files, args) : ''}
+        </clr-tree-node>`
```
Comment (kevinbuhmann) [inline .storybook/helpers/files.data.ts:95]:
```suggestion
      file => `
        <clr-tree-node
          [clrDisabled]="${file.disabled}"
          [clrExpanded]="${file.expanded}"
          ${args.clrSelected !== undefined && file.selected !== undefined ? `[clrSelected]="${file.selected}"` : ''}
        >
          ${args.asLink ? `<a href="javascript:;" class="clr-treenode-link">${file.name}</a>` : file.name}
          ${file.files ? getFileTreeNodeMarkup(file.files, args) : ''}
        </clr-tree-node>`
```

indentation is bit off

File: projects/angular/src/data/tree-view/models/tree-node.model.ts [hunk 15]
```diff
@@ -46,21 +47,38 @@ export abstract class TreeNodeModel<T> {
     this.selected.complete();
   }
 
+  toggleChildrenDisable() {
```
Comment (kevinbuhmann) [inline projects/angular/src/data/tree-view/models/tree-node.model.ts:50]: The word "toggle" in this method name implies that it flips the disabled state of the children from true to false or vice versa. It does not toggle the disabled state though. It just sets it match this node. I would rename this method to `propagateDisabledToChildren` or similar for clarity.

**Review by kevinbuhmann — CHANGES_REQUESTED** [review]

File: projects/angular/src/data/tree-view/models/tree-node.model.ts (hunk 15 above)
Comment (kevinbuhmann) [reply projects/angular/src/data/tree-view/models/tree-node.model.ts]: Like this: https://github.com/vmware-clarity/ng-clarity/commit/b081f58c05c90d731d650de9c9c7f83239fad678

Comment (kevinbuhmann) [reply projects/angular/src/data/tree-view/models/tree-node.model.ts]: You could also make the get accessor traverse up the tree (look at parent nodes) to see if there is a disabled ancestor node. I think there is an issue with the current code: You should be able to enable a node without enabling all of its children.

Comment (valentin-mladenov) [reply projects/angular/src/data/tree-view/models/tree-node.model.ts]: Disable functionality should work only towards children nodes. There is no use case where node should disable a parent, but all this is philosophical anyway and is not part of the current requirements.

Comment (valentin-mladenov) [reply projects/angular/src/data/tree-view/models/tree-node.model.ts]: Disabled accessor added in commit 64af64b

Comment (kevinbuhmann) [reply projects/angular/src/data/tree-view/models/tree-node.model.ts]: >There is no use case where node should disable a parent

I did not say that a node should disable a parent. I said a child node could check its parent nodes to see if it (the child node) is disabled.

File: projects/angular/src/data/tree-view/models/tree-node.model.spec.ts [hunk 16]
```diff
@@ -152,5 +152,20 @@ export default function (): void {
       child.toggleSelection(true);
       [...root.children, ...child.children].forEach(n => expect(n.selected.value).toBe(ClrSelectedState.UNSELECTED));
     });
+
+    it('Re enabled disabled root node do not change children disable status', function () {
```
Comment (kevinbuhmann) [inline projects/angular/src/data/tree-view/models/tree-node.model.spec.ts:156]:
```suggestion
    it('re enabled disabled root node do not change children disable status', function () {
```

nit

File: projects/angular/src/data/tree-view/models/tree-node.model.spec.ts [hunk 17]
```diff
@@ -152,5 +152,20 @@ export default function (): void {
...
+      child.disabled = true;
+
+      expect(root.disabled).toBe(false);
```
Comment (kevinbuhmann) [inline projects/angular/src/data/tree-view/models/tree-node.model.spec.ts:159]:
```suggestion
      expect(root.disabled).toBeFalse();
```

The code should be consistent.

Comment (valentin-mladenov) [reply projects/angular/src/data/tree-view/models/tree-node.model.spec.ts]: Corrected in commit 666bbec

File: projects/angular/src/data/tree-view/models/tree-node.model.spec.ts (hunk 16 above)
Comment (valentin-mladenov) [reply projects/angular/src/data/tree-view/models/tree-node.model.spec.ts]: Corrected in commit 666bbec

File: projects/angular/src/data/tree-view/tree-node.ts [hunk 18]
```diff
@@ -118,6 +119,14 @@ export class ClrTreeNode<T> implements OnInit, AfterContentInit, AfterViewInit,
     this._model.nodeId = this.nodeId;
   }
 
+  @Input('clrDisabled')
+  get disabled(): boolean {
+    return this._model.disabled;
+  }
+  set disabled(value: boolean) {
+    this._model.disabled = value || this._model.parent?.disabled;
```
Comment (kevinbuhmann) [inline projects/angular/src/data/tree-view/tree-node.ts:127]:
```suggestion
    this._model.disabled = value;
```

I don't think we need this anymore. We should write the state as given. The parent state is accounted for when read.

**Review by kevinbuhmann — CHANGES_REQUESTED** [review]

File: projects/demo/src/app/tree-view/eager-declarative-tree/eager-declarative-tree.html [hunk 19]
```diff
@@ -7,12 +7,12 @@
 <h4>No selection</h4>
 <div class="clr-example">
   <clr-tree>
-    <clr-tree-node>
+    <clr-tree-node [clrIfExpanded]="true">
```
Comment (kevinbuhmann) [inline projects/demo/src/app/tree-view/eager-declarative-tree/eager-declarative-tree.html:10]:
```suggestion
    <clr-tree-node [clrExpanded]="true">
```

File: projects/demo/src/app/tree-view/eager-declarative-tree/eager-declarative-tree.html [hunk 20]
```diff
@@ -7,12 +7,12 @@
...
       A
-      <clr-tree-node>A-1</clr-tree-node>
-      <clr-tree-node>
+      <clr-tree-node [clrDisabled]="true">A-1</clr-tree-node>
+      <clr-tree-node [clrIfExpanded]="true" [clrDisabled]="true">
```
Comment (kevinbuhmann) [inline projects/demo/src/app/tree-view/eager-declarative-tree/eager-declarative-tree.html:13]:
```suggestion
      <clr-tree-node [clrExpanded]="true" [clrDisabled]="true">
```

File: projects/angular/src/data/tree-view/tree-node.ts (hunk 18 above)
Comment (valentin-mladenov) [reply projects/angular/src/data/tree-view/tree-node.ts:127]: I thought I removed that.

File: projects/demo/src/app/tree-view/eager-declarative-tree/eager-declarative-tree.html (hunk 20 above)
Comment (valentin-mladenov) [reply projects/demo/src/app/tree-view/eager-declarative-tree/eager-declarative-tree.html]: Corrected in commit b5ee66f

File: projects/demo/src/app/tree-view/eager-declarative-tree/eager-declarative-tree.html (hunk 19 above)
Comment (valentin-mladenov) [reply projects/demo/src/app/tree-view/eager-declarative-tree/eager-declarative-tree.html]: Corrected in commit b5ee66f

**Review by kevinbuhmann — APPROVED** [review]
